### ISSUE TITLE
Adding specific norm methods for polynomials

### ISSUE THREAD

**Issue by blas-ko** [JuliaDiff/TaylorSeries.jl#127]
I feel that `abs(Taylor)` has a different approach as `norm(Taylor,p)` in this Package. `abs`here is a function 
abs: P → P, where P is a Polynomial
where `abs(P) = P` if `constant_term(P)` > 0 and `abs(P) = -P` if `constant_term(P)` < 0 

I think `norm(Taylor)` should stricly return a non-negative value such that
$$ ||P||_p=(sum_(k=0)^n|a_k|^p)^(1/p) $$

where ||P||_ ∞ = max_k |a_k| 

Right now
```julia
using TaylorSeries

X,Y = set_variables("x y");
norm(X-1,2)
  1.0 - 1.0 x + 𝒪(‖x‖⁷)
norm(X-1,Inf)
  1.0 - 1.0 x + 𝒪(‖x‖⁷)
```
so I propose extending `norm` for Taylors in this way, in a sense we always get a non-negative number as a result.

**Comment by PerezHz** [comment]
+1 for this idea! I think it is a good idea to be able to have a function, such as `norm`, which computes the [l-`p` norm of a polynomial](https://books.google.com.mx/books?id=wNTvBwAAQBAJ&lpg=PP1&pg=PA6#v=onepage&q&f=false). Actually, I have some work going in this direction; it relies on overloading `rtoldefault`, `isfinite`, and such functions from Julia `Base` for `Taylor1`, `HomogeneousPolynomial`, `TaylorN`, etc. This idea works at least on my machine, and would be happy to collaborate on a PR pursuing this.

Besides being related to #126, I think this kind of methods would make life much easier for comparing two polynomials, e.g., for testing in TaylorSeries and TaylorIntegration. 

Nevertheless, I understand there is a use case for `norm` as it is implemented now (@lbenet could clarify this), so we should think of being able to use both

**Comment by blas-ko** [comment]
I've made `norm` work as proposed, but some of `Base` functions had to be overwritten. The whole thing goes like this (this is using the `nested_taylor` branch from #126 ): 
```julia
using TaylorSeries
import TaylorSeries: NumberNotSeries

x,y = set_nested_variables(2,get_order())

import Base: norm, abs2
norm{T<:Number}(a::Taylor1{T},p::Real) = norm(a.coeffs,p)
norm{T<:Number}(a::TaylorN{T},p::Real) = norm(a.coeffs,p)
norm{T<:Number}(a::HomogeneousPolynomial{T},p::Real) = norm(a.coeffs,p)

function abs2{T<:NumberNotSeries}(a::Taylor1{T}) 
    a_coeffs = a.coeffs
    return dot(a_coeffs,a_coeffs)
end
function abs2{T<:NumberNotSeries}(a::HomogeneousPolynomial{T}) 
    a_coeffs = a.coeffs
    return dot(a_coeffs,a_coeffs)
end

#this are the overwritten functions; all of the norm(v) are changed for norm(v,p)
import Base.LinAlg: generic_vecnormp, vecnormInf, vecnormMinusInf, _length, vecnorm1,
                    generic_vecnorm1, generic_vecnormInf, generic_vecnormMinusInf
function generic_vecnormp(x, p)
    s = start(x)
    (v, s) = next(x, s)
    if p > 1 || p < -1 # might need to rescale to avoid overflow
        maxabs = p > 1 ? vecnormInf(x) : vecnormMinusInf(x)
        (maxabs == 0 || isinf(maxabs)) && return maxabs
        T = typeof(maxabs)
    else
        T = typeof(float(norm(v,p)))
    end
    spp::promote_type(Float64, T) = p
    if -1 <= p <= 1 || (isfinite(_length(x)*maxabs^spp) && maxabs^spp != 0) # scaling not necessary
        sum::promote_type(Float64, T) = norm(v,p)^spp
        while !done(x, s)
            (v, s) = next(x, s)
            sum += norm(v,p)^spp
        end
        return convert(T, sum^inv(spp))
    else # rescaling
        sum = (norm(v,p)/maxabs)^spp
        while !done(x, s)
            (v, s) = next(x, s)
            sum += (norm(v,p)/maxabs)^spp
        end
        return convert(T, maxabs*sum^inv(spp))
    end
end

function generic_vecnormInf(x)
    s = start(x)
    (v, s) = next(x, s)
    maxabs = norm(v,Inf)
    while !done(x, s)
        (v, s) = next(x, s)
        vnorm = norm(v,Inf)
        maxabs = ifelse(isnan(maxabs) | (maxabs > vnorm), maxabs, vnorm)
    end
    return float(maxabs)
end

function generic_vecnormMinusInf(x)
    s = start(x)
    (v, s) = next(x, s)
    minabs = norm(v,-Inf)
    while !done(x, s)
        (v, s) = next(x, s)
        vnorm = norm(v,-Inf)
        minabs = ifelse(isnan(minabs) | (minabs < vnorm), minabs, vnorm)
    end
    return float(minabs)
end

function generic_vecnorm1(x)
    s = start(x)
    (v, s) = next(x, s)
    av = float(norm(v,1))
    T = typeof(av)
    sum::promote_type(Float64, T) = av
    while !done(x, s)
        (v, s) = next(x, s)
        sum += norm(v,1)
    end
    return convert(T, sum)
end
```
This way, `norm` as the usual definition works as intended: 
```julia 
a = (3+4x)*y + 1 + x
  1.0 + 1.0 t + 𝒪(t⁷) +  3.0 + 4.0 t + 𝒪(t⁷) t + 𝒪(t⁷)
A = (3+4X)*Y + 1 + X
  1.0 + 1.0 x + 3.0 y + 4.0 x y + 𝒪(‖x‖⁷)

p = 2
@show norm(a,p)
@show norm(A,p)
@show norm([1,1,3,4],p)
@show (1^p+1^p+3^p+4^p)^(1./p)
  norm(a, p) = 5.196152422706632
  norm(A, p) = 5.196152422706632
  norm([1, 1, 3, 4], p) = 5.196152422706632
  (1 ^ p + 1 ^ p + 3 ^ p + 4 ^ p) ^ (1 ./ p) = 5.196152422706632

p = 3
@show norm(a,p)
@show norm(A,p)
@show norm([1,1,3,4],p)
@show (1^p+1^p+3^p+4^p)^(1./p)
  norm(a, p) = 4.530654896083492
  norm(A, p) = 4.530654896083492
  norm([1, 1, 3, 4], p) = 4.530654896083492
  (1 ^ p + 1 ^ p + 3 ^ p + 4 ^ p) ^ (1 ./ p) = 4.530654896083492

p = 1
@show norm(a,p)
@show norm(A,p)
@show norm([1,1,3,4],p)
@show (1^p+1^p+3^p+4^p)^(1./p)
  norm(a, p) = 9.0
  norm(A, p) = 9.0
  norm([1, 1, 3, 4], p) = 9.0
  (1 ^ p + 1 ^ p + 3 ^ p + 4 ^ p) ^ (1 ./ p) = 9.0

p = Inf
@show norm(a,p)
@show norm(A,p)
@show norm([1,1,3,4],p)
@show maximum([1,2,3,4])
  norm(a, p) = 4.0
  norm(A, p) = 4.0
  norm([1, 1, 3, 4], p) = 4.0
  maximum([1, 2, 3, 4]) = 4
```
Aditionaly, it's **fast**, so this is really promising for `TaylorIntegration`: 
```julia
p = 2
@time norm(a,p)
@time norm(A,p)
@time norm([1,1,3,4],p)
@time (1^p+1^p+3^p+4^p)^(1./p)
  0.000007 seconds (5 allocations: 176 bytes)
  0.000006 seconds (5 allocations: 176 bytes)
  0.000004 seconds (6 allocations: 288 bytes)
  0.002257 seconds (178 allocations: 11.840 KiB)
```
I want to PR this, but as I overwrite some internal functions for `norm` (especifically the `generic_norm` ones), I'm not sure what's the best way to proceed. This changes don't change the behaviour of normal `norm` methods with `AbstractArray`s or `Numbers`, so I think a PR in julia's file is the best thing to do for that functions. 

Anyways, what do you guys think? Sorry for the _really long_ comment!

**Comment by lbenet** [comment]
While I fully agree with the motivation (to have `norm` returning a non-negative number) and with the essence of the approach, I am not too sure that `abs2(x::AbstractSeries)` should be redefined.

Actually, I think your approach may be simplified. For instance:
```julia
julia> using TaylorSeries

julia> import Base: norm

julia> norm{T<:Number}(a::Taylor1{T},p::Real) = norm(a.coeffs,p) 
norm (generic function with 12 methods)

# NOTE the changes HERE
julia> norm{T<:Number}(a::TaylorN{T},p::Real) = norm([norm(x,p) for x in a.coeffs],p) 
norm (generic function with 13 methods)

julia> norm{T<:Number}(a::HomogeneousPolynomial{T},p::Real) = norm(a.coeffs,p)
norm (generic function with 14 methods)

# The following is for nested Taylor1's
julia> norm{T<:Number}(a::Taylor1{Taylor1{T}},p::Real) = norm([norm(x,p) for x in a.coeffs],p)
norm (generic function with 15 methods)

```

In the code above, I have only changed slightly the method `norm(a::TaylorN,p::Real)` and added a similar method for nested Taylor's. Then, using your definitions for the test cases:
```julia
julia> for p in (1,2,3,Inf,-Inf)
           @show(p)
           @show norm(a,p)
           @show norm(A,p)
           @show norm([1,1,3,4,0],p)
           println()
       end
p = 1
norm(a, p) = 9.0
norm(A, p) = 9.0
norm([1, 1, 3, 4, 0], p) = 9.0

p = 2
norm(a, p) = 5.196152422706632
norm(A, p) = 5.196152422706632
norm([1, 1, 3, 4, 0], p) = 5.196152422706632

p = 3
norm(a, p) = 4.530654896083492
norm(A, p) = 4.530654896083492
norm([1, 1, 3, 4, 0], p) = 4.530654896083492

p = Inf
norm(a, p) = 4.0
norm(A, p) = 4.0
norm([1, 1, 3, 4, 0], p) = 4.0

p = -Inf
norm(a, p) = 0.0
norm(A, p) = 0.0
norm([1, 1, 3, 4, 0], p) = 0.0
```

So everything matches!

Though I haven't checked the mixtures (`TaylorN{Taylor1{T}}` and the other way around), I think it may also work.

**Comment by lbenet** [comment]
Actually, the following definition may have all cases inside:
```julia
norm(x::TaylorSeries.AbstractSeries, p::Real) = norm([norm(a,p) for a in x.coeffs], p)
```
I haven't checked this thoroughly, but for the cases above, it works.

**Comment by blas-ko** [comment]
Great catch, @lbenet ! This works wonderfully! 
Actually, the best case I found is 
```julia 
norm(x::TaylorSeries.AbstractSeries, p::Real) = norm([norm(a,p) for a in x.coeffs], p)
norm{T<:Number}(a::HomogeneousPolynomial{T},p::Real) = norm(a.coeffs,p)
```
This works for more than 2 variables (I've tried until 4) and `Complex` coefficients!

**Comment by lbenet** [comment]
Great! Does it also work with mixtures (TaylorN{Taylor1{T}})? Can you please submit a PR? Thanks!

**Comment by lbenet** [comment]
It seems to me that mixtures must be treated separately:
```julia
julia> using TaylorSeries

julia> import Base: norm

julia> norm(x::TaylorSeries.AbstractSeries, p::Real) = norm([norm(a,p) for a in x.coeffs], p)
norm (generic function with 12 methods)

julia> norm(a::HomogeneousPolynomial, p::Real) = norm(a.coeffs,p)
norm (generic function with 13 methods)

julia> set_variables(Taylor1{Float64}, "x y")
2-element Array{TaylorSeries.TaylorN{TaylorSeries.Taylor1{Float64}},1}:
  ( 1.0 + 𝒪(t¹)) x + 𝒪(‖x‖⁷)
  ( 1.0 + 𝒪(t¹)) y + 𝒪(‖x‖⁷)

julia> norm(X)
ERROR: MethodError: no method matching abs2(::TaylorSeries.Taylor1{Float64})
Closest candidates are:
  abs2(::Bool) at bool.jl:78
  abs2(::Real) at number.jl:89
  abs2(::Complex) at complex.jl:216
  ...
```

However, with
```julia
norm{T<:Number}(x::HomogeneousPolynomial{T}, p::Real) = norm([norm(a,p) for a in x.coeffs], p)
```
at least the case above works.

**Comment by lbenet** [comment]
Incidentally, all methods above can be written simply as
```julia
norm(x::TaylorSeries.AbstractSeries, p::Real) = norm( norm.(x.coeffs, p), p)
```
since `HomogeneousPolynomial<:TaylorSeries.AbstractSeries` is `true`.

**Comment by PerezHz** [comment]
One thing I'd like to add (now that we're talking about `TaylorSeries.norm`), is that it'd be really convenient to have an `isapprox` method for `<:AbstractSeries` variables. This could be readily done, at least for julia 0.6+, by doing

```julia
using TaylorSeries
import Base: norm, rtoldefault, isfinite, isapprox

norm(x::AbstractSeries, p::Real=2) = norm( norm.(x.coeffs, p), p) #<--- awesome synthesis!

# rtoldefault and isfinite overloading for TaylorSeries
rtoldefault{T<:AbstractSeries}(::Type{T}) = sqrt(eps())
isfinite(x::AbstractSeries) = !isinf(x)

# the method below is an adaptation of Base.isapprox, MIT-licensed, as was taken from:
# https://github.com/JuliaLang/julia/blob/903644385b91ed8d95e5e3a5716c089dd1f1b08a/base/floatfuncs.jl#L203
function isapprox{T<:AbstractSeries}(x::T, y::T; rtol::Real=rtoldefault(x,y), atol::Real=0, nans::Bool=false)
    x == y || (isfinite(x) && isfinite(y) && norm(x-y,1) <= atol + rtol*max(norm(x,1), norm(y,1))) || (nans && isnan(x) && isnan(y))
end
```

which then may be used to do things like:

```julia

t = Taylor1(16)
s = sin(t)
x,y = set_nested_variables(2,get_order())
X, Y = set_variables(Taylor1{Float64}, "x y")

using Base.Test

@testset "Testing isapprox on next-to-trivial cases" begin
    @test t ≈ t
    @test t ≈ t+0.1sqrt(eps())
    @test s ≈ s
    @test s ≈ s+0.1sqrt(eps())
    @test x ≈ x
    @test x ≈ x+0.1sqrt(eps())
    @test y ≈ y
    @test y ≈ y+0.1sqrt(eps())
    @test X ≈ X
    @test X ≈ X+0.1sqrt(eps())
    @test Y ≈ Y
    @test Y ≈ Y+0.1sqrt(eps())
end
```

which works beautifully on julia 0.6+... problem is, with julia 0.5 this gives an ugly `MethodError` exception related to "no method matching `real(::Type{TaylorSeries.Taylor1{Float64}})`", which could be solved by defining the required `real` method, but I wonder if there's an easier way to do this?

**Comment by PerezHz** [comment]
(this could be *really* convenient for testing)

**Comment by PerezHz** [comment]
The other thing I noticed is that, at least for `Taylor1{Float64}` this unique method for `norm` is perhaps non-ideal wrt performance:

```julia
using TaylorSeries
import Base: norm, rtoldefault, isfinite, isapprox

norm(x::AbstractSeries, p::Real=2) = norm( norm.(x.coeffs, p), p) #<--- awesome synthesis!

# rtoldefault and isfinite overloading for TaylorSeries
rtoldefault{T<:AbstractSeries}(::Type{T}) = sqrt(eps())
isfinite(x::AbstractSeries) = !isinf(x)

# the method below is an adaptation of Base.isapprox, MIT-licensed, as was taken from:
# https://github.com/JuliaLang/julia/blob/903644385b91ed8d95e5e3a5716c089dd1f1b08a/base/floatfuncs.jl#L203
function isapprox{T<:AbstractSeries}(x::T, y::T; rtol::Real=rtoldefault(x,y), atol::Real=0, nans::Bool=false)
    x == y || (isfinite(x) && isfinite(y) && norm(x-y,1) <= atol + rtol*max(norm(x,1), norm(y,1))) || (nans && isnan(x) && isnan(y))
end

t = Taylor1(16)
s = sin(t)
```

which then on the REPL gives

```julia
julia> using BenchmarkTools

julia> @btime norm(s) #is there a way to improve performance here?
  104.014 ns (2 allocations: 240 bytes)
1.0138280236792707

julia> @btime norm(s.coeffs)
  76.441 ns (1 allocation: 16 bytes)
1.0138280236792707
```

So perhaps it'd be better to have separate methods of `norm` for specific cases such as `Taylor1{Float64}`, such as

```julia
norm(x::Taylor1, p::Real=2) = norm(x.coeffs, p)
```

which on the REPL yields

```julia
julia> @btime norm(s)
  68.590 ns (1 allocation: 16 bytes)
1.0138280236792707
```

**Comment by lbenet** [comment]
@PerezHz You are right; I think this has already been addressed in #128.

**Comment by PerezHz** [comment]
> @PerezHz You are right; I think this has already been addressed in #128.

Great!

**Comment by lbenet** [comment]
Solved by #128
